src/server/page_server.cpp

```diff
@@ -36,48 +36,46 @@ static void assert_page_server_type ();
 
 page_server::~page_server ()
 {
-  assert (m_ats_conn == nullptr);
   assert (m_replicator == nullptr);
+  assert (m_ats == nullptr);
 }
 
 void page_server::set_active_tran_server_connection (cubcomm::channel &&chn)
 {
   assert_page_server_type ();
 
   chn.set_channel_name ("ATS_PS_comm");
-  er_log_debug (ARG_FILE_LINE, "Active transaction server connected to this page server. Channel id: %s.\n",
-		chn.get_channel_id ().c_str ());
 
-  m_ats_conn.reset (new active_tran_server_conn (std::move (chn)));
-  m_ats_conn->register_request_handler (ats_to_ps_request::SEND_LOG_PRIOR_LIST,
-					std::bind (&page_server::receive_log_prior_list, std::ref (*this),
-					    std::placeholders::_1));
-  m_ats_conn->register_request_handler (ats_to_ps_request::SEND_LOG_PAGE_FETCH,
-					std::bind (&page_server::receive_log_page_fetch, std::ref (*this),
-					    std::placeholders::_1));
-  m_ats_conn->register_request_handler (ats_to_ps_request::SEND_DATA_PAGE_FETCH,
-					std::bind (&page_server::receive_data_page_fetch, std::ref (*this),
-					    std::placeholders::_1));
-
-  m_ats_conn->start_thread ();
-
-  m_ats_request_queue.reset (new active_tran_server_request_queue (*m_ats_conn));
-  m_ats_request_autosend.reset (new active_tran_server_request_autosend (*m_ats_request_queue));
-  m_ats_request_autosend->start_thread ();
+  const std::string channel_id { chn.get_channel_id () };
+  assert (m_ats == nullptr);
+  m_ats.reset (new ats_t (std::move (chn),
+  {
+    {
+      ats_to_ps_request::SEND_LOG_PRIOR_LIST,
+      std::bind (&page_server::receive_log_prior_list, std::ref (*this), std::placeholders::_1)
+    },
+    {
+      ats_to_ps_request::SEND_LOG_PAGE_FETCH,
+      std::bind (&page_server::receive_log_page_fetch, std::ref (*this), std::placeholders::_1)
+    },
+    {
+      ats_to_ps_request::SEND_DATA_PAGE_FETCH,
+      std::bind (&page_server::receive_data_page_fetch, std::ref (*this), std::placeholders::_1)
+    },
+  }));
+  er_log_debug (ARG_FILE_LINE, "PS connected to ATS. Channel id: %s.\n", channel_id);
```

Comment:
The connection is already created before m_ats is initialized. The message may be moved back and a copy of channel_id would not be required.

Reply:
Reverted.